File: sdk-typescript/src/protocol.test.ts
/**
 * `protocol.test.ts` — Tests du codec binaire, via le test runner intégré
 * de Node.js (`node --test`) pour ne pas imposer de dépendance de test
 * supplémentaire (Jest/Vitest) à un SDK destiné à être installé par des
 * tiers.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import {
  Opcode,
  ProtocolError,
  bytesToUuid,
  crc16CcittFalse,
  decodeFrame,
  encodeFrame,
  globMatch,
  uuidToBytes,
  FRAME_SIZE,
} from "./protocol.js";

const SAMPLE_TENANT = "12345678-9abc-def0-1122-334455667788";

test("encode puis decode restitue les mêmes champs", () => {
  const raw = encodeFrame({
    opcode: Opcode.Publish,
    tenantId: SAMPLE_TENANT,
    channelId: "room-42",
    payload: "hello world",
  });
  assert.equal(raw.length, FRAME_SIZE);

  const frame = decodeFrame(raw);
  assert.equal(frame.opcode, Opcode.Publish);
  assert.equal(frame.tenantId, SAMPLE_TENANT);
  assert.equal(frame.channelId, "room-42");
  assert.equal(frame.payload, "hello world");
});

test("uuidToBytes puis bytesToUuid fait un aller-retour identité", () => {
  const bytes = uuidToBytes(SAMPLE_TENANT);
  assert.equal(bytes.length, 16);
  assert.equal(bytesToUuid(bytes), SAMPLE_TENANT);
});

test("rejette une longueur de frame invalide", () => {
  assert.throws(() => decodeFrame(new Uint8Array(100)), ProtocolError);
});

test("rejette un magic invalide", () => {
  const raw = encodeFrame({ opcode: Opcode.Ping, tenantId: SAMPLE_TENANT });
  raw[0] = 0x00; // corrompt le magic
  assert.throws(() => decodeFrame(raw), ProtocolError);
});

test("rejette un CRC corrompu", () => {
  const raw = encodeFrame({ opcode: Opcode.Message, tenantId: SAMPLE_TENANT, payload: "data" });
  raw[50] = (raw[50] ?? 0) ^ 0xff; // altère un octet du payload sans mettre à jour le CRC
  assert.throws(() => decodeFrame(raw), ProtocolError);
});

test("tronque un channelId trop long sur une frontière UTF-8 valide", () => {
  const raw = encodeFrame({
    opcode: Opcode.Subscribe,
    tenantId: SAMPLE_TENANT,
    channelId: "x".repeat(100), // dépasse le champ de 24 octets
  });
  const frame = decodeFrame(raw);
  assert.equal(frame.channelId.length, 24);
});

test("crc16CcittFalse est déterministe et sensible à toute altération", () => {
  const a = new TextEncoder().encode("hello");
  const b = new TextEncoder().encode("hellp");
  assert.notEqual(crc16CcittFalse(a), crc16CcittFalse(b));
  assert.equal(crc16CcittFalse(a), crc16CcittFalse(a));
});

test("unsub opcode roundtrip", () => {
  const raw = encodeFrame({ opcode: Opcode.Unsub, tenantId: SAMPLE_TENANT, channelId: "room-42" });
  const frame = decodeFrame(raw);
  assert.equal(frame.opcode, Opcode.Unsub);
  assert.equal(frame.channelId, "room-42");
});

test("globMatch — motif simple avec wildcard final", () => {
  assert.ok(globMatch("orders:*", "orders:42"));
  assert.ok(!globMatch("orders:*", "invoices:42"));
});

test("globMatch — wildcard au milieu du motif", () => {
  assert.ok(globMatch("app_123:*:eu", "app_123:orders:eu"));
  assert.ok(!globMatch("app_123:*:eu", "app_123:orders:us"));
});

File: sdk-android/src/test/kotlin/com/yourorg/realtimesdk/ProtocolTest.kt
package com.yourorg.realtimesdk

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.UUID

/**
 * Tests du codec binaire (JUnit4, cohérent avec les conventions Android).
 *
 * ⚠️ N'a pas pu être exécuté dans l'environnement où ce SDK a été écrit —
 * ni `kotlinc` ni `javac` (JDK complet) n'y étaient disponibles, seul un
 * JRE. Ces tests suivent scrupuleusement les mêmes cas que
 * `protocol.test.ts` (TypeScript) et `test_protocol.py` (Python), qui
 * eux ont été réellement exécutés avec succès — la logique est donc
 * dérivée de code déjà validé, mais ce fichier précis reste à faire
 * tourner chez vous (`./gradlew test`) avant de le considérer fiable.
 */
class ProtocolTest {

    private val sampleTenant = UUID.fromString("12345678-9abc-def0-1122-334455667788")

    @Test
    fun `encode puis decode restitue les memes champs`() {
        val raw = encodeFrame(Opcode.PUBLISH, sampleTenant, "room-42", "hello world")
        assertEquals(FRAME_SIZE, raw.size)

        val frame = decodeFrame(raw)
        assertEquals(Opcode.PUBLISH, frame.opcode)
        assertEquals(sampleTenant, frame.tenantId)
        assertEquals("room-42", frame.channelId)
        assertEquals("hello world", frame.payload)
    }

    @Test
    fun `unicast opcode roundtrip`() {
        val raw = encodeFrame(Opcode.UNICAST, sampleTenant, "user-42", "hey there")
        val frame = decodeFrame(raw)
        assertEquals(Opcode.UNICAST, frame.opcode)
        assertEquals("user-42", frame.channelId)
    }

    @Test
    fun `unsub opcode roundtrip`() {
        val raw = encodeFrame(Opcode.UNSUB, sampleTenant, "room-42")
        val frame = decodeFrame(raw)
        assertEquals(Opcode.UNSUB, frame.opcode)
        assertEquals("room-42", frame.channelId)
    }

    @Test
    fun `rejette une longueur de frame invalide`() {
        assertThrows(ProtocolException::class.java) { decodeFrame(ByteArray(100)) }
    }

    @Test
    fun `rejette un magic invalide`() {
        val raw = encodeFrame(Opcode.PING, sampleTenant)
        raw[0] = 0x00 // corrompt le magic
        assertThrows(ProtocolException::class.java) { decodeFrame(raw) }
    }

    @Test
    fun `rejette un CRC corrompu`() {
        val raw = encodeFrame(Opcode.MESSAGE, sampleTenant, payload = "data")
        raw[50] = (raw[50].toInt() xor 0xFF).toByte() // altère un octet sans mettre à jour le CRC
        assertThrows(ProtocolException::class.java) { decodeFrame(raw) }
    }

    @Test
    fun `tronque un channelId trop long sur une frontiere UTF-8 valide`() {
        val raw = encodeFrame(Opcode.SUBSCRIBE, sampleTenant, channelId = "x".repeat(100))
        val frame = decodeFrame(raw)
        assertEquals(24, frame.channelId.length)
    }

    @Test
    fun `la troncature ne produit jamais d'UTF-8 invalide`() {
        // Emoji multi-octets (4 octets en UTF-8) répété pour forcer une
        // coupure au milieu d'un caractère si la troncature était naïve.
        val raw = encodeFrame(Opcode.PUBLISH, sampleTenant, payload = "\uD83C\uDF89".repeat(60))
        val frame = decodeFrame(raw) // ne doit lever aucune exception
        assertTrue(frame.payload.isNotEmpty())
    }

    @Test
    fun `crc16 est deterministe et sensible a toute alteration`() {
        val a = "hello".toByteArray(Charsets.UTF_8)
        val b = "hellp".toByteArray(Charsets.UTF_8)
        assertEquals(crc16CcittFalse(a), crc16CcittFalse(a))
        assertNotEquals(crc16CcittFalse(a), crc16CcittFalse(b))
    }

    @Test
    fun `globMatch avec wildcard final`() {
        assertTrue(globMatch("orders:*", "orders:42"))
        assertTrue(!globMatch("orders:*", "invoices:42"))
    }

    @Test
    fun `globMatch avec wildcard au milieu`() {
        assertTrue(globMatch("app_123:*:eu", "app_123:orders:eu"))
        assertTrue(!globMatch("app_123:*:eu", "app_123:orders:us"))
    }
}
